doc/developer/design/20241028_optimizer_releases.md

```diff
@@ -0,0 +1,81 @@
+# Optimizer Release Engineering
+
+## The Problem
+
+Currently, users run on "Materialize weekly", but:
+
+  1. That cannot be case forever, as self-hosted users may not upgrade immediately.
+  2. That is not always good for users, who can experience regressions when optimizer changes don't suit their workloads.
```

Comment:
> it will be much harder to pinpoint and fix regressions in a query if you need to look across six months of code changes rather than one week of code changes.

Yes, if there is 1 optimizer change in a given week, then it's immediately clear which change is to blame for a plan regression. With 6 months of changes, we'd need to do some non-trivial investigation. But I'm not too worried about this, because we have a powerful tool for this: Alexander's [optimizer trace tool](https://optimizer-trace.dev.materialize.com/) usually allows us to quickly (usually within 1-2 hours) figure out what's going on when an optimization is not going the way we imagined it. (Also note that I often need to run the trace tool anyway even when I know which PR is to blame, because it's often not clear how exactly the PR is breaking things, due to complex interactions between transforms.)

> I'm personally worried that releasing new versions of the optimizer every six months will shake out to _still_ result in too much breakage too frequently for users to be happy

If we have several months between getting to know about a regression and forcing a user to the regressing version, that gives us time to fix regressions:
- Sometimes another optimizer change can create an improvement for the same user, offsetting the regression, so it doesn't matter.
- Sometimes a regression is not inherently hard to fix once we know about the specific regression; maybe the regression happened just because we didn't think of some silly edge case. So, sometimes we'll be able to come up with an easy optimizer code fix.
- Sometimes it's a lot of work to fix a regression, but it can also align with our general optimizer evolution plans, so we'll do it anyway.
- Sometimes we'll opt to not fix a regression. Even in this case, there are still various other options for addressing the regression:
  - We might be able to manually rewrite the user's query to force a better plan. (This can involve chopping up the problematic part of the query to several indexed views, which are each small enough to not give much freedom to the optimizer to make bad decisions.)
  - We can compensate the user with credits, so they can simply scale up.
  - (We can introduce a feature flag tweaking some optimizer behavior for just that user. I hope to almost always avoid this.)